spot-contracts/contracts/RolloverVault.sol

```diff
@@ -565,33 +484,87 @@ contract RolloverVault is
     //--------------------------------------------------------------------------
     // Private write methods
 
-    /// @dev Deposits underlying balance into the provided bond and receives tranche tokens in return.
-    ///      And performs some book-keeping to keep track of the vault's assets.
-    /// @return balance The amount of underlying assets tranched.
-    /// @return bt The given bonds tranche data.
-    function _tranche(IBondController bond) private returns (uint256, BondTranches memory) {
-        // Get bond's tranche data
-        BondTranches memory bt = bond.getTranches();
+    /// @dev Redeems tranche tokens held by the vault, for underlying.
+    ///      NOTE: Reverts when attempting to recover a tranche which is not part of the deployed list.
+    ///      In the case of immature redemption, this method will recover other sibling tranches as well.
+    function _redeemTranche(ITranche tranche) private {
+        uint256 trancheBalance = tranche.balanceOf(address(this));
+        // if the vault has no tranche balance,
+        // we update our internal book-keeping and return.
+        if (trancheBalance <= 0) {
+            _syncAndRemoveDeployedAsset(tranche);
+            return;
+        }
 
-        // Get underlying balance
-        uint256 balance = underlying.balanceOf(address(this));
+        // get the parent bond
+        IBondController bond = IBondController(tranche.bond());
 
+        // if bond has matured, redeem the tranche token
+        if (bond.secondsToMaturity() <= 0) {
+            // execute redemption
+            _execMatureTrancheRedemption(bond, tranche, trancheBalance);
+
+            // sync deployed asset
+            _syncAndRemoveDeployedAsset(tranche);
+        }
+        // if not redeem using proportional balances
+        // redeems this tranche and it's siblings if the vault holds balances.
+        else {
+            // execute redemption
+            BondTranches memory bt = bond.getTranches();
+            _execImmatureTrancheRedemption(bond, bt);
+
+            // sync deployed asset, ie current tranche and all its siblings.
+            for (uint8 j = 0; j < bt.tranches.length; j++) {
+                _syncAndRemoveDeployedAsset(bt.tranches[j]);
+            }
+        }
+
+        // sync underlying
+        _syncAsset(underlying);
+    }
+
+    /// @dev Redeems perp tokens held by the vault for tranches.
+    function _redeemPerp(IPerpetualTranche perp_) private {
```

Comment:
Can this eventually be removed, once all the perps are redeemed and no more come in?

Guess it could still be used in future minting / swapping flows?

Reply:
Right this is essential for swapping flows